jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java

```diff
@@ -405,5 +409,76 @@ protected Selector newSelector() throws IOException
      */
     public abstract Connection newConnection(SelectableChannel channel, EndPoint endpoint, Object attachment) throws IOException;
 
+    public void addEventListener(EventListener listener)
+    {
+        if (isRunning())
+            throw new IllegalStateException(this.toString());
+        if (listener instanceof AcceptListener && !_acceptListeners.contains(listener))
+            _acceptListeners.add(AcceptListener.class.cast(listener));
+    }   
+    
+    public void removeEventListener(EventListener listener)
+    {
+        if (isRunning())
+            throw new IllegalStateException(this.toString());
+        if (listener instanceof AcceptListener)
+            _acceptListeners.remove(listener);
+    }
```

Comment:
Yuck. We may as well write Jetty in JavaScript, if we give up type safety in this way. I mean it works, but it's ugly. Why can't it be `addAcceptListener()` ? It will be impossible, for an embedded Jetty user to know what to pass to these methods without looking at their implementation. By using `AcceptListener` in the signature and name, it would be clear when the user types <ctrl+space> in the IDE.

Reply:
because the pattern is to have addEventListener(EventListener) to possibly support multiple listener types, potentially on the same object.    That is a pattern repeated elsewhere in the code.  Everytime we have deviated from that pattern it has caused trouble and confusion.   We should just follow the pattern.

Reply:
I agree with @sbordet

Reply:
@sbordet @joakime just because HttpClient didn't follow an established pattern when it probably should have, is not a reason to do so again.

We have lots of listeners in jetty and more are probably coming.  We do not want to have to write code like:
```
MyFooBarListener listener = new MyFooBarListener();
o.addFooListener(listener);
o.addBarListener(listener);
```
Currently the pattern is that we can simply write:
```
o.addEventListener(new MyFooBarListener());
```
or in some circumstances:

```
o.addBean(new MyFooBarListener());
```

Also, most of our interfaces have multiple methods, so lambas are not even possible. For those that might have a single method it is possible to use lambas with:
```
o.addEventListener((FooListener)e->{});
```

Changing the design if code so that your particular IDE might be able to work better is just entirely the wrong way around.   But if you really REALLY need to do that, then it is always possible to add an intermediary interface that extends EventListener, so that your IDE can offer a list of types:
```
class SomeThing
{
    interface SomeThingEventListener extends EventListener{};
    public void addEventListener(SomeThingEventListener listener);
}
```

Finally, having a single method is not motivated by having a single backing data structure. In fact it is more efficient to [split out the listeners into separate lists](https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L673-L694). Note also how this is easily extended to that [derived classes can add new listener types](https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java#L188-L201) without adding new API to the class.

So you have to give me better reasons than these for us to break this pattern.  If anything, I think we need to find and fix some of the examples where [this pattern has been broken which do cause confusion](https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-util/src/main/java/org/eclipse/jetty/util/component/AbstractLifeCycle.java#L139)!

Reply:
https://github.com/eclipse/jetty.project/issues/2578

Reply:
Ugh! The more I look at this, the more it is apparent that most of the listeners we have added recently have all decided to invent their own pattern of being set: via adding beans, via dedicated APIs , via discovery!

It is a bit of a mess and we need to re-establish a core pattern, perhaps with a few well documented deviations: #2578